fedimint-server/src/net/queue.rs

```diff
@@ -60,49 +84,209 @@ where
         }
+
+        // If we got an ACK from the future we remove all of them from the queue even if we haven't
+        // sent them yet (can happen on reconnect). In that case we need to adjust the unsent
+        // messages field so we never try to send messages not in the buffer anymore
+        let queued_messages = self.queue.len() as u64;
+        if self.unsent_messages > queued_messages {
+            self.unsent_messages = queued_messages;
+        }
+    }
+
+    /// Fetch the next message to be sent from the queue
+    pub fn next_send_message(&self) -> Option<&UniqueMessage<M>> {
+        if self.unsent_messages > 0 {
+            let idx = self.queue.len() - (self.unsent_messages as usize);
+            // If we always properly update unsent_messages this can never fail
+            let maybe_message = self
+                .queue
+                .get(idx)
+                .expect("We tried to send a message no longer in the buffer");
+
+            Some(maybe_message)
+        } else {
+            None
+        }
     }
 
-    pub fn iter(&self) -> impl Iterator<Item = &UniqueMessage<M>> {
-        self.queue.iter()
+    /// Marks the message `msg_id` as sent over the wire. This does not aknowledge the receipt by
+    /// our peer and thus keeps the message in the buffer in case it needs to be re-sent.
+    pub fn mark_sent(&mut self, msg_id: MessageId) {
+        assert_ne!(self.unsent_messages, 0, "There are no messages to be sent!");
+        assert_eq!(
+            msg_id.0,
+            self.next_id.0 - self.unsent_messages,
+            "The sent message is not the expected one!"
+        );
+        self.unsent_messages -= 1;
+    }
+
+    /// Mark all messages as unsent to attempt re-sending and return the oldest and newest messages
+    /// to be sent.
+    pub fn resend_all(&mut self) -> Option<(MessageId, MessageId)> {
+        self.unsent_messages = self.queue.len() as u64;
+
+        if self.queue.is_empty() {
+            None
+        } else {
+            let oldest = self.queue.front().expect("Queue not empty").id;
+            let newest = self.queue.back().expect("Queue not empty").id;
+            Some((oldest, newest))
+        }
+    }
+
+    /// Return the number of unsent messages
+    pub fn unsent_len(&self) -> usize {
+        self.unsent_messages as usize
+    }
+}
+
+impl<M: MaybeEpochMessage> MessageQueue<M> {
+    /// Returns the number of epochs the buffer has messages of. This assumes that the we send
+    /// messages in each epoch and the epoch only increases.
+    pub fn buffered_epochs(&self) -> u64 {
+        // We only have non-epoch messages during DKG and on re-joins. Both happen infrequently
+        // enough that this O(n) search for the lowest and highest epoch number in the buffer is
+        // sufficient
+        let maybe_oldest = self
+            .queue
+            .iter()
+            .filter_map(|msg| msg.msg.message_epoch())
+            .next();
```

Comment:
is clippy working in CI?
it should be complaining about filter_map.next => find_map

Reply:
fixed